File: server/src/graphql_api/modules/user/mutations.py
import uuid

import strawberry
from strawberry.experimental import pydantic as strawberry_pydantic
from strawberry.types import Info

from core.users.dependencies import get_user_service
from core.users.schema import UserCreate, UserUpdate
from graphql_api.helpers import IsSuperAdmin
from graphql_api.modules.user.types import UserType


@strawberry_pydantic.input(model=UserCreate, all_fields=False)
class UserCreateInput:
    identifier: str = strawberry.UNSET
    description: str = ""
    password: str | None = None


@strawberry_pydantic.input(model=UserUpdate, all_fields=True)
class UserUpdateInput:
    password: str | None = None


@strawberry.type
class UserMutation:
    @strawberry.mutation(permission_classes=[IsSuperAdmin])
    async def create_user(self, info: Info, input: UserCreateInput) -> UserType:
        requester = info.context["request"].state.user
        session = info.context["session"]
        service = get_user_service(session=session)
        return await service.create_user(user=input.to_pydantic(), requester=requester)

    @strawberry.mutation(permission_classes=[IsSuperAdmin])
    async def update_user(self, info: Info, id: uuid.UUID, body: UserUpdateInput) -> UserType:
        requester = info.context["request"].state.user

        session = info.context["session"]
        service = get_user_service(session=session)
        user_update = body.to_pydantic()
        return await service.update_user(user_id=id, user=user_update, requester=requester)

    @strawberry.mutation(permission_classes=[IsSuperAdmin])
    async def link_user_account(self, info: Info, primary_user_id: uuid.UUID, secondary_user_id: uuid.UUID) -> UserType:
        requester = info.context["request"].state.user

        session = info.context["session"]
        service = get_user_service(session=session)
        return await service.link_accounts(
            primary_user_id=primary_user_id, secondary_user_id=secondary_user_id, requester=requester
        )

    @strawberry.mutation(permission_classes=[IsSuperAdmin])
    async def unlink_user_account(
        self, info: Info, primary_user_id: uuid.UUID, secondary_user_id: uuid.UUID
    ) -> UserType:
        requester = info.context["request"].state.user
        session = info.context["session"]
        service = get_user_service(session=session)
        return await service.unlink_accounts(
            primary_user_id=primary_user_id, secondary_user_id=secondary_user_id, requester=requester
        )

File: ui/frontend-lib/src/users/graphql/mutations.ts
export const CREATE_USER_MUTATION = `
  mutation CreateUser($input: UserCreateInput!) {
    createUser(input: $input) {
      id
      identifier
      entityName
    }
  }
`;

/**
 * Partial payload for updating a single user field at a time, used by the
 * inline editing controls on the user overview page.
 *
 * Field names match the backend `UserUpdateInput` GraphQL type.
 */
export type UserUpdateFieldInput = Partial<{
  description: string;
  deactivated: boolean;
  password: string;
}>;

export const UPDATE_USER_MUTATION = `
  mutation UpdateUser($id: UUID!, $body: UserUpdateInput!) {
    updateUser(id: $id, body: $body) {
      id
      identifier
      entityName
    }
  }
`;

export const LINK_USER_ACCOUNT_MUTATION = `
  mutation LinkUserAccount($primaryUserId: UUID!, $secondaryUserId: UUID!) {
    linkUserAccount(
      primaryUserId: $primaryUserId
      secondaryUserId: $secondaryUserId
    ) {
      id
      identifier
      entityName
    }
  }
`;

export const UNLINK_USER_ACCOUNT_MUTATION = `
  mutation UnlinkUserAccount($primaryUserId: UUID!, $secondaryUserId: UUID!) {
    unlinkUserAccount(
      primaryUserId: $primaryUserId
      secondaryUserId: $secondaryUserId
    ) {
      id
      identifier
      entityName
    }
  }
`;
